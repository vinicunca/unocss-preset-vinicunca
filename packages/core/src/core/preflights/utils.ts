export function compressCSS(css: string) {
  return css.replace(/\s+/g, ' ').replace(/\/\*[\s\S]*?\*\//g, '');
}
