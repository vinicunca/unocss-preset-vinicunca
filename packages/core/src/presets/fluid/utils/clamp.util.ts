import type { RequiredFluidOptions } from '../../../types';

import { extractValuesFromRegexMatch } from './regex.util';
import { getRemMaxWidth, getRemMinWidth } from './rem.util';

interface GetSlopeParams {
  min: number;
  max: number;
  config: RequiredFluidOptions;
}

/**
 * Calculates the slope value for fluid.
 *
 * @param options - GetSlopeParams.
 * @param options.min - The minimum value in the fluid range.
 * @param options.max - The maximum value in the fluid range.
 * @param options.config - Configuration object for additional settings.
 *
 * @returns The calculated slope value.
 */
export function getSlope({ min, max, config }: GetSlopeParams) {
  const remMaxWidth = getRemMaxWidth(config);
  const remMinWidth = getRemMinWidth(config);
  return (
    (max - min) / (remMaxWidth - remMinWidth)
  );
}

interface GetIntersectionParams {
  min: number;
  slope: number;
  minWidth: number;
}

/**
 * Calculates the intersection point for the fluid formula.
 *
 * @param options - GetIntersectionParams.
 * @param options.min - The minimum value in the fluid range.
 * @param options.slope - The slope value calculated for the fluid.
 * @param options.minWidth - The minimum viewport width.
 *
 * @returns The calculated intersection point.
 */
export function getIntersection({ min, slope, minWidth }: GetIntersectionParams) {
  return (-minWidth * slope + min);
}

interface GetSlopePercentageParams {
  min: number;
  max: number;
  config: RequiredFluidOptions;
}

/**
 * Calculates the slope percentage for the CSS 'clamp' function.
 *
 * @param options - GetSlopePercentageParams.
 * @param options.min - The minimum value in the fluid range.
 * @param options.max - The maximum value in the fluid range.
 * @param options.config - Configuration object.
 *
 * @returns The slope value as a percentage for use in the 'clamp' function.
 */
export function getSlopePercentage({ min, max, config }: GetSlopePercentageParams) {
  const slope = getSlope({ min, max, config });
  return (slope * 100);
}

interface GetClampParams {
  min: number;
  max: number;
  config: RequiredFluidOptions;
}

/**
 * Generates a CSS 'clamp' function value for fluid.
 *
 * @param options - GetClampParams.
 * @param options.min - The minimum value in the fluid range.
 * @param options.max - The maximum value in the fluid range.
 * @param options.config - Configuration object.
 *
 * @returns A string representing the CSS 'clamp' function.
 */
export function getClamp({ min, max, config }: GetClampParams) {
  const slope = getSlope({ min, max, config });
  const intersection = getIntersection({ min, slope, minWidth: getRemMinWidth(config) });
  const slopePercentage = getSlopePercentage({ min, max, config });

  const clampMin = Math.min(min, max);
  const clampMax = Math.max(min, max);

  return `clamp(${clampMin}rem, ${intersection.toFixed(4)}rem + ${slopePercentage.toFixed(4)}vw, ${clampMax}rem)`;
}

interface GetClampCommentParams {
  match: RegExpMatchArray;
  config: RequiredFluidOptions;
}

/**
 * Generates a comment with the fluid values for the clamp CSS.
 *
 * @param options - GetClampCommentParams.
 * @param options.match - The match array from the regex.
 * @param options.config - Configuration object.
 *
 * @returns A string representing the comment for the 'clamp' function with fluid real values.
 */
export function getClampComment({ match, config }: GetClampCommentParams) {
  if (!config.commentHelpers) {
    return '';
  }

  const { predefinedRangeName, matchMin, matchMax } = extractValuesFromRegexMatch(match);
  const predefinedRange = config.ranges && config.ranges[predefinedRangeName!];
  const isRem = config.useRemByDefault;
  const min = predefinedRange ? predefinedRange[0] : -Number.parseInt(matchMin!);
  const max = predefinedRange ? predefinedRange[1] : -Number.parseInt(matchMax!);
  const unit = isRem ? 'rem' : 'px';

  return `/* ${min}${unit} -> ${max}${unit} */`;
}
