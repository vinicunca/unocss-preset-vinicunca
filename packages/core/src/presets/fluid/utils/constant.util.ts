export const FLUID_UTILITIES = {
  'fluid-text': 'font-size',
  'fluid-w': 'width',
  'fluid-min-w': 'min-width',
  'fluid-max-w': 'max-width',
  'fluid-h': 'height',
  'fluid-min-h': 'min-height',
  'fluid-max-h': 'max-height',
  'fluid-p': 'padding',
  'fluid-pt': 'padding-top',
  'fluid-pb': 'padding-bottom',
  'fluid-pl': 'padding-left',
  'fluid-pr': 'padding-right',
  'fluid-px': ['padding-left', 'padding-right'],
  'fluid-py': ['padding-top', 'padding-bottom'],
  'fluid-m': 'margin',
  'fluid-mt': 'margin-top',
  'fluid-mb': 'margin-bottom',
  'fluid-ml': 'margin-left',
  'fluid-mr': 'margin-right',
  'fluid-mx': ['margin-left', 'margin-right'],
  'fluid-my': ['margin-top', 'margin-bottom'],
  'fluid-gap': 'gap',
  'fluid-gap-x': 'column-gap',
  'fluid-gap-y': 'row-gap',
  'fluid-indent': 'text-indent',
  'fluid-scroll-m': 'scroll-margin',
  'fluid-scroll-mt': 'scroll-margin-top',
  'fluid-scroll-mb': 'scroll-margin-bottom',
  'fluid-scroll-ml': 'scroll-margin-left',
  'fluid-scroll-mr': 'scroll-margin-right',
  'fluid-scroll-mx': ['scroll-margin-left', 'scroll-margin-right'],
  'fluid-scroll-my': ['scroll-margin-top', 'scroll-margin-bottom'],
  'fluid-scroll-ms': 'scroll-margin-inline-start',
  'fluid-scroll-me': 'scroll-margin-inline-end',
  'fluid-scroll-p': 'scroll-padding',
  'fluid-scroll-pt': 'scroll-padding-top',
  'fluid-scroll-pb': 'scroll-padding-bottom',
  'fluid-scroll-pl': 'scroll-padding-left',
  'fluid-scroll-pr': 'scroll-padding-right',
  'fluid-scroll-px': ['scroll-padding-left', 'scroll-padding-right'],
  'fluid-scroll-py': ['scroll-padding-top', 'scroll-padding-bottom'],
  'fluid-scroll-ps': 'scroll-padding-inline-start',
  'fluid-scroll-pe': 'scroll-padding-inline-end',
  'fluid-leading': 'line-height',
  'fluid-top': 'top',
  'fluid-right': 'right',
  'fluid-bottom': 'bottom',
  'fluid-left': 'left',
  'fluid-inset': 'inset',
  'fluid-inset-x': ['left', 'right'],
  'fluid-inset-y': ['top', 'bottom'],
};
