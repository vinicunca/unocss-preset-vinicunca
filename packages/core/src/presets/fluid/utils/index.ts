export * from './clamp.util';
export * from './constant.util';
export * from './parser.util';
export * from './regex.util';
export * from './rem.util';
export * from './rule.util';
export * from './validation.util';
