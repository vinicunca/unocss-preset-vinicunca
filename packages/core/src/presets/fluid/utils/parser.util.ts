/*
 * Casts a string to a number, if possible.
 * @param value - The string to cast.
 * @returns The cast number, or 0 if the string is empty.
 */
export function invertAndParseNumber(value?: string) {
  if (!value) {
    return 0;
  }

  if (value.includes('-')) {
    return Number.parseInt(value.replace('-', ''));
  }

  return -Number.parseInt(value);
}
