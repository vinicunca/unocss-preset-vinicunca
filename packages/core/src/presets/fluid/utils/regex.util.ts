export const REGEX_PATTERNS_NUMERIC_VALUES = '(?:--)?(-?\\d+)?(?:--)?(-?\\d+)?$';
export const REGEX_PATTERNS_RANGE_VALUES = '(?:--)?(-?\\d+)?(?:-([a-zA-Z0-9]+))?$';

export function extractValuesFromRegexMatch(match: RegExpMatchArray) {
  const [utility, matchMin, matchMax, , predefinedRangeName] = match;

  return {
    utility,
    matchMin,
    matchMax,
    predefinedRangeName,
  };
}
