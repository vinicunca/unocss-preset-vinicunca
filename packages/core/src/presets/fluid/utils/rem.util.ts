import type { RequiredFluidOptions } from '../../../types';

import { invertAndParseNumber } from './parser.util';
import { validateUtilityRange } from './validation.util';

export function getRemMaxWidth(config: RequiredFluidOptions) {
  const maxWidth = config.extendMaxWidth || config.maxWidth;

  if (config.useRemByDefault) {
    return maxWidth;
  } else {
    return maxWidth / config.remBase;
  }
}

export function getRemMinWidth(config: RequiredFluidOptions) {
  const minWidth = config.extendMinWidth || config.minWidth;
  if (config.useRemByDefault) {
    return minWidth;
  } else {
    return minWidth / config.remBase;
  }
}

export function toRem(value: number, config: RequiredFluidOptions) {
  if (config.useRemByDefault) {
    return value;
  } else {
    return value / config.remBase;
  }
}

/**
 * Calculates the relative size based on the original viewport size and the new viewport size.
 * This allows you to extend the min and max sizes of the fluid layout keeping the proportions.
 * This function is used for the extendMinWidth and extendMaxWidth options.
 * The result of this value is used to calculate the min and max values in rem when the viewport are extended.
 */
export function calculateRelativeSize(
  {
    originalViewPortMin,
    originalMinSize,
    originalMaxSize,
    originalViewPortMax,
    newViewPortSize,
  }:
  {
    originalViewPortMin: number;
    originalMinSize: number;
    originalMaxSize: number;
    originalViewPortMax: number;
    newViewPortSize: number;
  },
) {
  const slope = (originalMaxSize - originalMinSize)
    / (originalViewPortMax - originalViewPortMin);
  return slope * (newViewPortSize - originalViewPortMin) + originalMinSize;
}

interface ExtractRemBoundsFromMatchParams {
  match: RegExpMatchArray | string;
  config: RequiredFluidOptions;
}

/**
 * Returns the min and max values in rem for the given match.
 */
export function extractRemBoundsFromMatch({ match, config }: ExtractRemBoundsFromMatchParams) {
  let max, min;

  const [utility, matchMin, matchMax, , predefinedRangeName] = match;

  if (predefinedRangeName) {
    validateUtilityRange({ range: predefinedRangeName, utility, config });
    const [minRange, maxRange] = config.ranges![predefinedRangeName]!;
    min = toRem(minRange, config);
    max = toRem(maxRange, config);
  } else {
    min = toRem(invertAndParseNumber(matchMin), config);
    max = toRem(invertAndParseNumber(matchMax), config);
  }

  let relativeMin: number | undefined;
  let relativeMax: number | undefined;

  if (config.extendMinWidth) {
    relativeMin = calculateRelativeSize(
      {
        originalViewPortMin: config.minWidth,
        originalMinSize: min,
        originalMaxSize: max,
        originalViewPortMax: config.maxWidth,
        newViewPortSize: config.extendMinWidth,
      },
    );
  }

  if (config.extendMaxWidth) {
    relativeMax = calculateRelativeSize(
      {
        originalViewPortMin: config.minWidth,
        originalMinSize: min,
        originalMaxSize: max,
        originalViewPortMax: config.maxWidth,
        newViewPortSize: config.extendMinWidth!,
      },
    );
  }

  return {
    min: relativeMin ?? min,
    max: relativeMax ?? max,
  };
}
