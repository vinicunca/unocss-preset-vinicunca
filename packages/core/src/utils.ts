import type { CSSObject } from '@unocss/core';
import type { ThemeAnimation } from '@unocss/preset-mini';
import type { CustomStaticShortcuts } from './types';

// name duration timing-function iteration-count
export const RE_ANIMATION = /^([a-z-]+)\s+([0-9.]+m?s?|[*+])?\s?([a-z-]+(?:\([^)]+\))?|[*+])?\s*([a-z0-9-]+|[*+])?$/i;

/**
 * Normalize custom animate usage to UnoCSS animations theme.
 *
 * ⚠️ We must strictly follow the following format. ⚠️
 *
 * [ name, duration, timing-function, iteration-count ]
 *
 * If you use * as placeholder, it will be ignored.
 *
 * [name, duration, *, iteration-count]
 *
 * If you use + as placeholder, it will be replaced with empty string.
 *
 * [name, duration, +, iteration-count]
 *
 * @example
 *
 * { animate: ['spin 1s linear infinite'] }
 *
 * Will be transformed:
 *
 * {
 *   animate: ['spin 1s linear infinite'],
 *   durations: {
 *     spin: '1s',
 *   },
 *   timingFns: {
 *     spin: 'linear',
 *   },
 *   counts: {
 *     spin: 'infinite',
 *   },
 * }
 */
export function resolveAnimation(extendAnimation: Record<string, string>) {
  const animation: ThemeAnimation = {};
  const keys: Array<
    Exclude<keyof ThemeAnimation, 'properties'>
  > = ['durations', 'timingFns', 'counts'];
  const shortcuts: CustomStaticShortcuts = [];

  // TODO: replace for in with for of to remove the no-restricted-syntax eslint rule
  // eslint-disable-next-line no-restricted-syntax
  for (const key_ in extendAnimation) {
    const val = extendAnimation[key_];
    const match = val.match(RE_ANIMATION);

    if (match != null) {
      const [, name, duration, timing, count] = match;
      const values = [duration, timing, count];

      if (name !== key_) {
        shortcuts.push([`animate-${key_}`, `animate-${name}`]);
      }

      for (let i = 0; i < keys.length; i++) {
        const key = keys[i];
        const value = values[i];
        if (value != null) {
          if (value === '*') {
            continue;
          }

          if (animation[key]) {
            animation[key]![name] = values[i] === '+' ? '' : values[i];
          } else {
            animation[key] = {
              [name]: values[i] === '+' ? '' : values[i],
            };
          }
        }
      }
    }
  }

  return {
    animation,
    shortcuts,
  };
}

export function cssObj2StrSync(style: Record<string, CSSObject>): string {
  return Object.keys(style).reduce((str, key) => {
    return `${str}${key}${stringifyObj(style[key])}`;
  }, '').replace(/\n/g, '');
}

export function stringifyObj(obj: CSSObject) {
  return `{${Object.keys(obj).reduce((str, key) => {
    return `${str}${camelToHyphen(key)}:${obj[key]};`;
  }, '')}}`;
}

function camelToHyphen(str: string): string {
  return str.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
}

export function compressCSS(css: string) {
  return css.replace(/\s+/g, ' ').replace(/\/\*[\s\S]*?\*\//g, '');
}
