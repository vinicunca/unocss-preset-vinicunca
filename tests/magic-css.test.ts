import { expect, it } from 'vitest';

import { getCssFromUnoWithoutPreflights } from './utils';

it('preset-animatecss', async () => {
  const { css } = await getCssFromUnoWithoutPreflights({
    options: {
      magicCss: true,
    },
    code: 'animate-magic',
  });

  expect(css).toMatchSnapshot();
});
