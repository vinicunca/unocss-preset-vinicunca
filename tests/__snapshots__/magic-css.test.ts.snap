// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preset-animatecss 1`] = `
"/* layer: preflights */
*,::before,::after{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgb(0 0 0 / 0);--un-ring-shadow:0 0 rgb(0 0 0 / 0);--un-shadow-inset: ;--un-shadow:0 0 rgb(0 0 0 / 0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgb(147 197 253 / 0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}::backdrop{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-pan-x: ;--un-pan-y: ;--un-pinch-zoom: ;--un-scroll-snap-strictness:proximity;--un-ordinal: ;--un-slashed-zero: ;--un-numeric-figure: ;--un-numeric-spacing: ;--un-numeric-fraction: ;--un-border-spacing-x:0;--un-border-spacing-y:0;--un-ring-offset-shadow:0 0 rgb(0 0 0 / 0);--un-ring-shadow:0 0 rgb(0 0 0 / 0);--un-shadow-inset: ;--un-shadow:0 0 rgb(0 0 0 / 0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgb(147 197 253 / 0.5);--un-blur: ;--un-brightness: ;--un-contrast: ;--un-drop-shadow: ;--un-grayscale: ;--un-hue-rotate: ;--un-invert: ;--un-saturate: ;--un-sepia: ;--un-backdrop-blur: ;--un-backdrop-brightness: ;--un-backdrop-contrast: ;--un-backdrop-grayscale: ;--un-backdrop-hue-rotate: ;--un-backdrop-invert: ;--un-backdrop-opacity: ;--un-backdrop-saturate: ;--un-backdrop-sepia: ;}
/* layer: default */
@keyframes magic{0% {opacity: 1;-webkit-transform-origin: 100% 200%;transform-origin: 100% 200%;-webkit-transform: scale(1, 1) rotate(0deg);transform: scale(1, 1) rotate(0deg);}100% {opacity: 0;-webkit-transform-origin: 200% 500%;transform-origin: 200% 500%;-webkit-transform: scale(0, 0) rotate(270deg);transform: scale(0, 0) rotate(270deg);}}
.animate-magic{animation:magic 1s linear 1;animation-fill-mode:both;}"
`;
